state/execution.go

```diff
@@ -221,10 +218,24 @@ func (blockExec *BlockExecutor) CreateProposalBlock(
 	// TODO: make sure that the hashes are correct here
 	// via also removing hashes that the application removed!
 	hashes := make([][]byte, len(newData.Txs))
-	for i := 0; i < len(newData.Txs); i++ {
-		hashes[i] = newData.Txs[i].Hash()
+	numWorkers := runtime.NumCPU()
+	if numWorkers > len(newData.Txs) {
+		numWorkers = len(newData.Txs)
+	}
+	workers := make(chan struct{}, numWorkers)
+	var wg sync.WaitGroup
+	for i, tx := range newData.Txs {
+		wg.Add(1)
+		go func() {
+			workers <- struct{}{}
```

Comment:
Writing to channel before spawning routine prevents overspawning of routines. Currently the code will spawn a routine for every transaction, while you need only numWorkers amount of them.

Reply:
check now: b1444d0a629f6c02345e090e323b87516186ef50 this should only spin up routines up to `max(cpu count, number of transactions)` which is what we want

Reply:
Why do you need tx amount of routines? Whats the goal?

Reply:
my bad, I meant `min(cpu count, number of transactions)`

Reply:
Oh, i might have read that if case wrong. Changing it to builtin `min` could help

Reply:
> to see how much ms we could shave off on high load, if we reuse routines

That makes sense. I went from the assumption that go routines are pretty lightweight to spawn. Also, during this period, nothing intensive is happening in the app, that's why I decided to just go for the simple approach. 
Would be nice to have a way to benchmark both tho

Reply:
this is already giving us ~100ms per block with spinning up new go routines. I think reusing go routines can be the next step that we keep in mind if we need to squeeze more performance. But as you said, we have a few other places where we can get performance.

For nmt/rsmt2d, Mikhail, who will join our team shortly, has put up really good work to parallelize, re-use buffers, etc, in nmt and rsmt2d to squeeze more performance: 
- https://github.com/celestiaorg/nmt/pull/304
- https://github.com/celestiaorg/rsmt2d/pull/379
Since the change is core to our protocol, it would be amazing if you could also review it as you're good with performance stuff.

Reply:
sure, take your time. The reason why it's exposed is to have the choice whether to use it or not. The downside of using it is if you want to generate a proof, it won't work. That's why both APIs are exposed.